platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/counters/BackpressureObjectCounter.java

```diff
@@ -72,38 +73,43 @@ public BackpressureObjectCounter(
-            //
-            // This is logically equivalent to the following pseudocode.
-            // Note that the managed block is thread safe when onRamp() is being called from multiple threads,
-            // even though this pseudocode is not.
-            //
-            //                while (count >= capacity) {
-            //                    Thread.sleep(sleepNanos);
-            //                }
-            //                count++;
-            //
-            // The reason why we use the managedBlock() strategy instead of something simpler has to do with
-            // the fork join pool paradigm. Unlike traditional thread pools where we have more threads than
-            // CPUs, blocking (e.g. Thread.sleep()) on a fork join pool may monopolize an entire CPU core.
-            // The managedBlock() pattern allows us to block while yielding the physical CPU core to other
-            // tasks.
-            ForkJoinPool.managedBlock(onRampBlocker);
-        } catch (final InterruptedException ex) {
-            // This should be impossible.
-            Thread.currentThread().interrupt();
-            throw new IllegalStateException("Interrupted while blocking on an onRamp() for " + name);
+            // Slow case. Capacity wasn't reserved, so we may need to block.
+
+            try {
+                // This will block until capacity is available and the count has been incremented.
+                //
+                // This is logically equivalent to the following pseudocode.
+                // Note that the managed block is thread safe when onRamp() is being called from multiple threads,
+                // even though this pseudocode is not.
+                //
+                //                while (count >= capacity) {
+                //                    Thread.sleep(sleepNanos);
+                //                }
+                //                count++;
+                //
+                // The reason why we use the managedBlock() strategy instead of something simpler has to do with
+                // the fork join pool paradigm. Unlike traditional thread pools where we have more threads than
+                // CPUs, blocking (e.g. Thread.sleep()) on a fork join pool may monopolize an entire CPU core.
+                // The managedBlock() pattern allows us to block while yielding the physical CPU core to other
+                // tasks.
+                ForkJoinPool.managedBlock(onRampBlocker);
+                return;
+            } catch (final InterruptedException ex) {
+                // This should be impossible.
+                Thread.currentThread().interrupt();
+                throw new IllegalStateException("Interrupted while blocking on an onRamp() for " + name);
+            } catch (final RejectedExecutionException ex) {
+                // We've exhausted our supply of background threads, we have no choice but to busy wait.
```

Comment:
Should we sleep a bit?

Reply:
Possibly. @OlegMazurov what would you recommend that we do inside this loop?

Reply:
Add `Thread.onSpinWait()`. Instantiating and throwing the exception is expensive enough to pass for a delay between unsuccessful attempts. Will be clearly visible in profiles (we'll reassess if and when that happens).
The real issue is that we use back-pressure inside a ForkJoinPool.

Reply:
done